modules/search-as-you-type/src/main/java/org/elasticsearch/index/mapper/SearchAsYouTypeFieldMapper.java

```diff
@@ -0,0 +1,432 @@
+import org.apache.lucene.analysis.shingle.FixedShingleFilter;
+import org.apache.lucene.document.Field;
+import org.apache.lucene.index.IndexOptions;
+import org.apache.lucene.index.IndexableField;
+import org.apache.lucene.index.Term;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.TermQuery;
+import org.elasticsearch.common.collect.Iterators;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.support.XContentMapValues;
+import org.elasticsearch.index.analysis.AnalyzerScope;
+import org.elasticsearch.index.analysis.NamedAnalyzer;
+import org.elasticsearch.index.query.QueryShardContext;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.index.mapper.TypeParsers.nodeIndexOptionValue;
+import static org.elasticsearch.index.mapper.TypeParsers.parseMultiField;
+
+public class SearchAsYouTypeFieldMapper extends FieldMapper implements ArrayValueMapperParser {
+
+    private static final Logger LOG = LogManager.getLogger(SearchAsYouTypeFieldMapper.class);
+
+    public static class Defaults {
+
+        public static final int MIN_GRAM = 1;
+        public static final int MAX_GRAM = 20;
+        public static final int MAX_SHINGLE_SIZE = 3;
+
+        public static final MappedFieldType FIELD_TYPE = new SearchAsYouTypeFieldType();
+
+        static {
+            FIELD_TYPE.setOmitNorms(true);
+            FIELD_TYPE.setIndexOptions(IndexOptions.DOCS_AND_FREQS_AND_POSITIONS);
+            FIELD_TYPE.freeze();
+        }
+    }
+
+    public static class TypeParser implements Mapper.TypeParser {
+
+        @Override
+        public Mapper.Builder<?, ?> parse(String name,
+                                          Map<String, Object> node,
+                                          ParserContext parserContext) throws MapperParsingException {
+
+            final SearchAsYouTypeFieldMapper.Builder builder = new SearchAsYouTypeFieldMapper.Builder(name);
+
+            NamedAnalyzer analyzer = parserContext.getIndexAnalyzers().getDefaultIndexAnalyzer();
+            int maxShingleSize = Defaults.MAX_SHINGLE_SIZE;
+
+            for (Iterator<Map.Entry<String, Object>> iterator = node.entrySet().iterator(); iterator.hasNext();) {
+                final Map.Entry<String, Object> entry = iterator.next();
+                final String fieldName = entry.getKey();
+                final Object fieldNode = entry.getValue();
+                if (fieldName.equals("type")) {
+                    continue;
+                } else if (fieldName.equals("index_options")) {
+                    builder.indexOptions(nodeIndexOptionValue(fieldNode));
+                    iterator.remove();
+                } else if (fieldName.equals("fields")) {
```

Comment:
We should not handle multi fields here, we already have sub-fields so no need to add more complexity ?

Reply:
Strongly agree